kedro/io/data_catalog.py

```diff
@@ -105,28 +105,36 @@ def __init__(
         """Return a _FrozenDatasets instance from some datasets collections.
         Each collection could either be another _FrozenDatasets or a dictionary.
         """
+        self._original_names: set[str] = set()
         for collection in datasets_collections:
             if isinstance(collection, _FrozenDatasets):
                 self.__dict__.update(collection.__dict__)
+                self._original_names.update(collection._original_names)
             else:
                 # Non-word characters in dataset names are replaced with `__`
                 # for easy access to transcoded/prefixed datasets.
-                self.__dict__.update(
-                    {
-                        _sub_nonword_chars(dataset_name): dataset
-                        for dataset_name, dataset in collection.items()
-                    }
-                )
+                for dataset_name, dataset in collection.items():
+                    self.__dict__[_sub_nonword_chars(dataset_name)] = dataset
+                    self._original_names.add(dataset_name)
 
     # Don't allow users to add/change attributes on the fly
     def __setattr__(self, key: str, value: Any) -> None:
+        if key == "_original_names":
+            super().__setattr__(key, value)
+            return
         msg = "Operation not allowed! "
         if key in self.__dict__:
             msg += "Please change datasets through configuration."
         else:
             msg += "Please use DataCatalog.add() instead."
         raise AttributeError(msg)
 
+    def _ipython_key_completions_(self) -> list[str]:
+        return list(self._original_names)
```

Comment:
So there is an argument that simplicity may be better, but what about including the `dataset.__class__.__name__` so you would see something like `shuttles (CSVDataset)`

Reply:
I would also artificially place `parameters` at the very bottom to discourage its use

Reply:
> So there is an argument that simplicity may be better, but what about including the `dataset.__class__.__name__` so you would see something like `shuttles (CSVDataset)`?

I think it's a good idea and I see the value. The only concern is whether we are fine with those keys. Cause with tab completion we will get `catalog.datasets["shuttles (CSVDataset)"]` which will work but not sure it looks good from the syntax side.

Reply:
I'll check if it's possible to provide some metadata which is not part of the key but still displayed with `IPython` and `Jupyter` UI.